homeassistant/components/sonos/media.py

```diff
@@ -0,0 +1,235 @@
+"""Support for media metadata handling."""
+from __future__ import annotations
+
+import datetime
+import logging
+from typing import Any
+
+from soco.core import (
+    MUSIC_SRC_AIRPLAY,
+    MUSIC_SRC_LINE_IN,
+    MUSIC_SRC_RADIO,
+    MUSIC_SRC_SPOTIFY_CONNECT,
+    MUSIC_SRC_TV,
+    SoCo,
+)
+from soco.data_structures import DidlAudioBroadcast, DidlPlaylistContainer
+from soco.music_library import MusicLibrary
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import dispatcher_send
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    SONOS_MEDIA_UPDATED,
+    SONOS_STATE_PLAYING,
+    SONOS_STATE_TRANSITIONING,
+    SOURCE_AIRPLAY,
+    SOURCE_LINEIN,
+    SOURCE_SPOTIFY_CONNECT,
+    SOURCE_TV,
+)
+from .helpers import soco_error
+
+LINEIN_SOURCES = (MUSIC_SRC_TV, MUSIC_SRC_LINE_IN)
+SOURCE_MAPPING = {
+    MUSIC_SRC_AIRPLAY: SOURCE_AIRPLAY,
+    MUSIC_SRC_TV: SOURCE_TV,
+    MUSIC_SRC_LINE_IN: SOURCE_LINEIN,
+    MUSIC_SRC_SPOTIFY_CONNECT: SOURCE_SPOTIFY_CONNECT,
+}
+UNAVAILABLE_VALUES = {"", "NOT_IMPLEMENTED", None}
+DURATION_SECONDS = "duration_in_s"
+POSITION_SECONDS = "position_in_s"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _timespan_secs(timespan: str | None) -> None | float:
+    """Parse a time-span into number of seconds."""
+    if timespan in UNAVAILABLE_VALUES:
+        return None
+
+    assert timespan is not None
+    return sum(60 ** x[0] * int(x[1]) for x in enumerate(reversed(timespan.split(":"))))
```

Comment:
Very smart, very unreadable 😝

Reply:
This does something similar, returns a timedelta instead.

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/config_validation.py#L424

Reply:
This has been there since the [very first `sonos` commit](https://github.com/home-assistant/core/commit/1b874c603b666888f829d583d36ed57505fc49d3#diff-e3102b464276703e90884c4b464313af1e28fbd41e4e56d7215714311db969f2R106-R107) 😛 

Never a bad time to improve readability, done in https://github.com/home-assistant/core/pull/66840/commits/0b633a46c88a433069283b24ded58d1ccdf0982f 👍